Fetch the GitHub members

IonicaBizau opened the pull request:
This pull request adds the functionality to use the public GitHub API to fetch the GitHub members. :star2: 

To access the GitHub API, I used [`gh.js`](https://github.com/IonicaBizau/gh.js)–a library created by me. Here is how the code looks like:

``` js
var gh = new GitHub();
gh.get("orgs/github/members", { all: true }, function (err, members) {
   /* do something with these awesome people */
});
```

Obviously, after three games (there are 3 requests to the members endpoint and 16 requests to `users/<login>` per game–and the rate limit is 60) it will fallback to the `js/Hubbers.js` information, due to the public rate limit. Why 16? Because there are 16 squares and because the `name` resource doesn't appear in the `members` endpoint responses. :balloon: 

To update the information in that file, I built a tiny script (`build.js`) which uses `gh.js` on the server, fetches the needed info and saves it into the `Hubbers.js` file. This way we can easily update the members info. :rocket: 

``` sh
$ node build.js <token>
```

@alysonla Please test it out and see what else should be fixed/improved. :grin: Of course, the docs should be updated too.

And yeah, it fixes #5! :four_leaf_clover:

IonicaBizau committed:
Include gh.js on the page

IonicaBizau committed:
Removed script tag which was not opened

IonicaBizau committed:
Fixed the favicon

IonicaBizau committed:
Updated gh.min.js

IonicaBizau committed:
Fetch the hubbers from GitHub API

IonicaBizau committed:
Updated gh.min.js

IonicaBizau committed:
Added gitignore–so my VIM temp files will be ignored :grin:

IonicaBizau committed:
Use avatar_url instead of avatar

IonicaBizau committed:
Create new requests to get the names

IonicaBizau committed:
Added the build script

Since the orgs/github/members endpoint does not give us name information, we have to query the name for each member. That means a lot of requests. This script will do it for us, but it needs a token (since there are many requests). So the usage is:

node build.js <token>

It will override the js/Hubbers.js file with the new information. On the client side, when the requests fail (due to the rate limit), it will fallback to this file information which contains all the members of the GitHub organization (fetched when the script was run).

IonicaBizau committed:
Removed the timeout

IonicaBizau commented:
@alysonla And should we add a visual thing that shows to the user that something is loading? :thought_balloon:

alysonla commented:
@IonicaBizau Amazing!!! :sparkling_heart:  Testing this out now!

IonicaBizau commented on js/memory-matching.js:
```diff
@@ -1,19 +1,27 @@
-console.log(matchingGame.deck)
+function shuffleHubbers(hubbers) {
+    // Like this https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/slice
+    hubbers.hubbers.sort(shuffle).slice(0,8).map(function(hubber){
```
@alysonla This is not really random. :laughing: It's cute, but not really good solution to shuffle elements. See [this comment](http://stackoverflow.com/questions/2450954/how-to-randomize-shuffle-a-javascript-array#comment29711767_18650169):

> Downvoting as this isn't really that random. I don't know why it has so many upvotes. Do not use this method. It looks pretty, but isn't completely correct. Here are results after 10,000 iterations on how many times each number in your array hits index [0](I can give the other results too): 1 = 29.19%, 2 = 29.53%, 3 = 20.06%, 4 = 11.91%, 5 = 5.99%, 6 = 3.32% 

This happens for an array like `[1, 2, 3, 4, 5, 6]`–so, in our case (~200 GitHub public members), the first ones (you are one of them, since they are sorted alphabetically probably :smile: ) will have the greatest chance to appear in the blocks, while the last ones will have the lowest chance to appear in the game. Black magic math! :trollface:

alysonla replied:
Oh interesting. Based on the link, it sounds like the Fisher-Yates shuffle algorithm would be the best/most correct implementation, yeah?

IonicaBizau replied:
Yep, and [here](http://bost.ocks.org/mike/shuffle/) we have a nice visualization. :bar_chart:

alysonla commented on build.js:
```diff
@@ -0,0 +1,57 @@
+
+// Fetch GitHub members
+Logger.log("Fetching the GitHub members.");
+gh.get("orgs/github/members", { all: true }, function (err, data) {
```
@IonicaBizau is this where the get request is fetching all the github org members? Sadly I know nothing about how to use the GitHub API. :smile:

alysonla replied:
Oh it looks like you addressed this in your opening PR comment. 


``` js
var gh = new GitHub();
gh.get("orgs/github/members", { all: true }, function (err, members) {
   /* do something with these awesome people */
});
```

IonicaBizau replied:
> Sadly I know nothing about how to use the GitHub API.

Ha! GitHub API is probably the best API I have ever used. Everything is so natural. `{ all: true}` tells [`gh.js`](https://github.com/IonicaBizau/gh.js) to handle pagination automatically (max is 100 per page), so more requests are done in the background. :sparkle:

alysonla commented on index.html:
```diff
@@ -1,7 +1,7 @@
 <html lang="en">
 	<head>
-		<link rel="icon" type="image/ico" href="http://alysonla.github.io/hubber_memory_game/favicon.ico"/>
+		<link rel="icon" type="image/ico" href="favicon.ico"/>
```
Thanks for cleaning this up :+1:

IonicaBizau replied:
:grin:

alysonla commented on build.js:
```diff
@@ -0,0 +1,57 @@
+  , SameTime = require("same-time")
+  , Logger = require("bug-killer")
+  , Fs = require("fs")
+  ;
```
@IonicaBizau maybe a noob question but where are these dependencies pulling in from?

IonicaBizau replied:
@alysonla From [`npm`](https://www.npmjs.com/). To install them, we have two options:
- create a [`package.json`](https://docs.npmjs.com/files/package.json) file and have there the `dependencies`
- `npm install` them manually: `npm i gh.js same-time bug-killer`

(`fs` is Node.js built-in module)

I would choose the first one.

alysonla commented:
@IonicaBizau Thanks so much for this PR. SO EXCITING!

After testing it out, I noticed that the full name is pulling in on both card tiles instead of the name on one and the handle on the other, as it was previously. Would you be able to fix that? 

> To update the information in that file, I built a tiny script (build.js) which uses gh.js on the server, fetches the needed info and saves it into the Hubbers.js file. This way we can easily update the members info.

Oh this is really nice, thank you! 

I'm also wondering what it might take for this be a "fork & go" type project - i.e. this could be used for any GitHub organization (up to 3 times, it sounds like) - in theory, the build.js & hubbers.js files could be removed and the get request would just be replaced with the name of the new Org. For example if the Twitter org wanted to use it, it'd be as simple as changing a couple lines - the title, h1 header, & 

``` js
var gh = new GitHub();
gh.get("orgs/twitter/members", { all: true }, function (err, members) {
   /* do something with these awesome people */
});
```

As I'm typing this, I realize I can test this out myself so gonna try it now - fingers crossed! :grinning:

alysonla commented:
Oh it worked! The code to change is in the memory-matching.js file 

``` js
 function getHubbers(callback) {
   matchingGame.github.get("orgs/twitter/members", { all: true }, function (err, data) {
        if (err) {
            console.warn(err);
            data = window.Hubbers;
        }
        callback(null, { hubbers: data });
    });
```

Super easy! 

As you noted the rate limit will kick in after 3 attempts however it's sounds like it resets after 1 hour at least. https://developer.github.com/v3/#rate-limiting

IonicaBizau committed:
Display usernames and names

@alysonla Since we need to make additional requests for real names, we need to do that in the getHubbers name, otherwise there will be no element displaying the real name when the blocks are rendering (because of the API requests).

Now, we make sure that the names are there and we display them like before. :sparkle:

IonicaBizau commented:
> I'm also wondering what it might take for this be a "fork & go" type project - i.e. this could be used for any GitHub organization (up to 3 times, it sounds like) - in theory, the build.js & hubbers.js files could be removed and the get request would just be replaced with the name of the new Org. For example if the Twitter org wanted to use it, it'd be as simple as changing a couple lines - the title, h1 header, &

Oh, yeah! It could even be a querystring parameter–e.g. http://alysonla.github.io/hubber-memory-game/?org=twitter

The rate limit could be fixed using a token on the client side, provided by the user, knowing we are _not_ evil. :smile:

> After testing it out, I noticed that the full name is pulling in on both card tiles instead of the name on one and the handle on the other, as it was previously. Would you be able to fix that?

Addressed in https://github.com/IonicaBizauKitchen/hubber-memory-game/commit/218a72786a70b0bbd2800cfd26f018d5c3a7a135. :sparkles:

IonicaBizau committed:
Use the Fisher-Yates (aka Knuth) algorithm to shuffle the hubbers

IonicaBizau commented on js/memory-matching.js:
```diff
@@ -1,22 +1,60 @@
 
-Hubbers["hubbers"].sort(shuffle).slice(0,8).map(function(hubber){
-	matchingGame.deck.push(hubber)
-	matchingGame.deck.push(hubber)
```
@alysonla Wowow, always learning something: this can be written as `matchingGame.deck.push(hubber, hubber)`. :laughing:

IonicaBizau committed:
Optimized the number of API requests

Only make requests for the 8 users we want to display.

IonicaBizau commented:
...and improved in https://github.com/IonicaBizauKitchen/hubber-memory-game/commit/42b22031266425c622063e0c9b954a4e64807430 :balloon: :tada: 

![image](https://cloud.githubusercontent.com/assets/2864371/10264628/58b4a096-6a1a-11e5-91ba-5bb812e86cfb.png)

IonicaBizau committed:
Spaces instead of tabs

IonicaBizau committed:
Spaces instead of tabs

IonicaBizau committed:
Added the CSS spinner

IonicaBizau commented:
Added a CSS spinner from [here](http://www.css-spinners.com/spinner/three-quarters). Looks better now! :smile:

![image](https://cloud.githubusercontent.com/assets/2864371/10264701/7c92f66e-6a1c-11e5-929f-d4bfdd74c6b1.png)

IonicaBizau committed:
New GitHubbers!

alysonla commented:
@IonicaBizau thanks so much - this all looks great AND I learned a lot. :pray:  I'm going to update the readme & clean up some other files in another PR. 

> Oh, yeah! It could even be a querystring parameter–e.g. http://alysonla.github.io/hubber-memory-game/?org=twitter

Also this would be neat!

IonicaBizau commented:
@alysonla Great! :tada: You're welcome! You just need to wait until Francisc grows up a little bit and then he can have fun with this game! :baby: :smile:

> Also this would be neat!

Maybe, that could be a different project using my library. :smile:

IonicaBizau commented:
@alysonla  ...and it's so funny when common faces appear when the blocks are flipped. :smile:

IonicaBizau commented:
>  I'm going to update the readme & clean up some other files in another PR.

@alysonla I did that for you. :memo: See #10.
